File: 1201-1300/1261-Find Elements in a Contaminated Binary Tree/1261-Find Elements in a Contaminated Binary Tree.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class FindElements:

    def __init__(self, root: TreeNode):
        self.table = set()
        def dfs(root, curr):
            if root is None:
                return
            root.val = curr
            self.table.add(curr)
            dfs(root.left, curr * 2 + 1)
            dfs(root.right, curr * 2 + 2)
            
        dfs(root, 0)

    def find(self, target: int) -> bool:
        return target in self.table


# Your FindElements object will be instantiated and called as such:
# obj = FindElements(root)
# param_1 = obj.find(target)

File: 1201-1300/1261-Find Elements in a Contaminated Binary Tree/1261-Find Elements in a Contaminated Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class FindElements {
public:
    FindElements(TreeNode* root) {
        dfs(root, 0);
    }
    
    bool find(int target) {
        return table.count(target);
    }
    
private:
    unordered_set<int> table;
    void dfs(TreeNode* root, int curr) {
        if (root == nullptr) {
            return;
        }
        
        root->val = curr;
        table.insert(curr);
        dfs(root->left, curr * 2 + 1);
        dfs(root->right, curr * 2 + 2);
    }
};

/**
 * Your FindElements object will be instantiated and called as such:
 * FindElements* obj = new FindElements(root);
 * bool param_1 = obj->find(target);
 */
